File: tarot-backend/src/main/kotlin/data/tarot/Card.kt
package xclavel.data.tarot

import xclavel.data.server.Player

data class Card(
    val id: Int,
    val value: Int,
    val color: Color,
    var owner: Player? = null,
    val status: CardStatus = CardStatus.IN_HAND,
) {

    companion object {
        fun fromId(id: Int, owner: Player?) : Card {
            val card = when (id) {
                in 1..14 -> Card(id, id, Color.CARREAU)
                in 15..28 -> Card(id, id - 14, Color.TREFLE)
                in 29..42 -> Card(id, id - 28, Color.COEUR)
                in 43..56 -> Card(id, id - 42, Color.PIQUE)
                in 57..77 -> Card(id, id - 56, Color.ATOUT)
                78 -> Card(id, -1, Color.EXCUSE)
                else -> throw Exception("Invalid card id")
            }
            card.owner = owner
            return card
        }
    }



    fun countPoint(): Float {
        return if (isOudler()) {
            4.5f //bouts
        } else if (color == Color.ATOUT) {
             0.5f
        } else {
            when (value) {
                11 -> 1.5f //valet
                12 -> 2.5f //cavalier
                13 -> 3.5f //dame
                14 -> 4.5f //roi
                else -> 0.5f //carte quelconque
            }
        }
    }

    fun isOudler() =
        color == Color.EXCUSE ||
        color == Color.ATOUT && (
            value == 1 ||
            value == 21
        )

    fun isValet() = value == 11
    fun isCavalier() = value == 12
    fun isDame() = value == 13
    fun isRoi() = value == 14

    fun isExcuse() = value == -1

    fun isRegularCard() = color != Color.ATOUT && color != Color.EXCUSE
}

File: tarot-game/Assets/Scripts/Card.cs
public class Card
{
    public readonly int index;
    public readonly int value;
    public readonly TarotColor color;

    public bool isRoi() => value == 14;
    public bool isReine() => value == 13;
    public bool isCavalier() => value == 12;
    public bool isValet() => value == 11;
    public bool isExcuse() => value == -1;

    public string toString()
    {
        if (isExcuse()) return "Excuse";
        if (color == TarotColor.ATOUT) return $"{value} d'atout";
        if (isRoi()) return $"Roi de {color}";
        if (isReine()) return $"Reine de {color}";
        if (isCavalier()) return $"Cavalier de {color}";
        if (isValet()) return $"Valet de {color}";
        return $"{value} de {color}";
    }
    
    public Card(int index)
    {
        this.index = index;
        switch (index)
        {
            case <= 14:
                this.color = TarotColor.CARREAU;
                this.value = index;
                break;
            
            case <= 28:
                this.color = TarotColor.TREFLE;
                this.value = index - 14;
                break;
            
            case <= 42:
                this.color = TarotColor.COEUR;
                this.value = index - 28;
                break;
            
            case <= 56:
                this.color = TarotColor.PIQUE;
                this.value = index - 42;
                break;
            
            case <= 77:
                this.color = TarotColor.ATOUT;
                this.value = index - 56;
                break;
            
            case 78:
                this.color = TarotColor.EXCUSE;
                this.value = -1;
                break;
        }
    }
}
